### DISCUSSION TITLE
OAuth2 with Github always returns `oauth2: server response missing access_token`

### DISCUSSION THREAD

**Discussion by CedricVanhaverbeke** [pocketbase/pocketbase#611] [Q&A] [answered]
I tried OAuth2 with Github. I easily configured a client id and client secret with the Pocketbase admin ui. But when i try the `client.users.authViaOAuth2` method, I always get a `400` status code with error (in the Pocketbase backend  oauth2: server response missing access_token).

It seems the error originates here: https://github.com/golang/oauth2/blob/5d25da1a8d43b66f2898c444f899c7bcfd6a407e/internal/token.go#L282

I just followed the tutorial with a small `React` application. Is there something I'm doing wrong here?

**Comment by ganigeorgiev** [comment]
Hm, are you passing the code from the redirect response to the `authViaOAuth2` call?

I'm not sure if this is what you mean by tutorial, but you could find a PocketBase OAuth2 example in the "Auth via OAuth2" section in https://pocketbase.io/docs/manage-users/#auth-via-oauth2.

_I'll move the issue to Q&A discussion._

**Comment by CedricVanhaverbeke** [comment, marked as answer]
For some reason, it suddenly just works. Perhaps some weird caching issue.

Thanks for the quick feedback! :D

**Comment by vpavlin** [comment]
I just stumbled upon this discussion when searching for the error in PB logs:

```
Status: 400
Method: POST
Auth: guest
URL: /api/collections/users/auth-with-oauth2
Referer: http://localhost:3000/

{
  "errorMessage": "oauth2: server response missing access_token"
}
```

It is the last part of the oauth flow - I got the list of providers, selected Github, got redirected to GH auth, approved, got redirected back and then got

```
ClientResponseError 400: Something went wrong while processing your request.
    at new ClientResponseError (pocketbase.es.mjs?696c:1:2274)
    at eval (pocketbase.es.mjs?696c:1:41012)
 ```
 
 in frontend and the above is the corresponding PB logs entry.

**Reply by ganigeorgiev** [reply]
I'm not able to reproduce it.

Based on the `Referer` header, do you have a middleware service that could interrupt with the redirect?

Could you try to reproduce it with just the 2 static files as shown in https://pocketbase.io/docs/authentication/#web-oauth2-integration?

**Reply by vpavlin** [reply]
I am thinking whether a problem might be one of the following

a) PB running in cloud (something.defidoc.org)
b) FE running locally (localhost:3000)


I'll try with the static files

**Reply by vpavlin** [reply]
Call me tired or stupid, but I cannot get the `pb_public` serving working - is there anything extra I need to do apart from having that folder with index.html there? I get 

```
{
code: 404,
message: "Not Found.",
data: { }
}
```

When I access the PB url

**Reply by ganigeorgiev** [reply]
Sorry I forgot to mention, if you use PocketBase as framework (aka. have custom build), then you have to register the public static files route manually, something like:

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    e.Router.GET("/*", apis.StaticDirectoryHandler(os.DirFS("./pb_public"), false))
    return nil
})
```

**Reply by vpavlin** [reply]
Ok, that works:) So the question is, what am I missing:D

**Reply by vpavlin** [reply]
:facepalm: My redirect URL was misconfigured in `authWithOAuth2` call - i.e. there was a mismatch between that one and the one configured in GH Oauth App - sadly that error did not surface anywhere

**Reply by vpavlin** [reply]
ONe note for the doc - would it make sense to use something like this: 

```
const pb = new PocketBase(location.protocol + '//' + location.host);
const redirectUrl = location.protocol + '//' + location.host + '/redirect.html';
```

Instead of hardcoding `127.0.0.1:8090` in the example? Or is the goal to "force" people to edit those files if their values differ to make them realize the actual values?:D

**Reply by ganigeorgiev** [reply]
I think it is better to have the static literal so that is more clear what actually is happening. The OAuht2 docs also could be used as a reference by non-JS developers since we don't have currently integration examples for other platforms and could be confusing at first glance if you are not familiar with the browser apis.

But in general, the OAuth2 will be simplified anyway in the near future (v0.15/v0.16) as part of #55 and a user provided redirect handler will no longer be required since ideally everything will happen with a single `authWithOAuth2` call.

**Reply by vpavlin** [reply]
Fair enough, thanks for your patience:)

**Reply by vpavlin** [reply]
Ok...premature excitement - when I logged out, it stopped working again:D

This is basically my next.js based `redirect.html`
```javascript
import { usePocketBase } from '@/hooks/usePocketBase';
import { useRouter } from 'next/router';
import { RecordAuthResponse } from 'pocketbase';
import { useEffect, useState } from 'react';

const Callback = () => {
    const { pb } = usePocketBase();
    const [authData, setAuthData] = useState<RecordAuthResponse>();

    const router = useRouter();
    const { state, code } = router.query;
    console.log(router.query);

    useEffect(() => {
        if (!state || !code) return;
        const providerStr = localStorage.getItem('provider');

        if (providerStr == null) return;
        const provider = JSON.parse(providerStr);

        console.log(state, code.toString(), provider);
        // compare the redirect's state param and the stored provider's one
        if (provider.state !== state) {
            throw "State parameters don't match.";
        }

        if (!code) throw 'Failed to get code';

        pb.collection('users')
            .authWithOAuth2(
                provider.name,
                code,
                provider.codeVerifier,
                'http://localhost:3000/callback`',
                // pass optional user create data
                {
                    emailVisibility: false,
                }
            )
            .then((authData) => {
                console.log(authData);
                setAuthData(authData);
            })
            .catch((e) => console.log(e));
    }, [state, code]);

    // authenticate

    return <>Auth Data: {authData?.token}</>;
};

export default Callback;
```

The code and state are loaded from URL params successfully (I can see them in console and state matches the on in provider data)

**Reply by vpavlin** [reply]
double :facepalm:  - I was gonna delete it, but I am going to leave it here as a memento - check the redirect url parameter - somehow an extra backtick made it's way in there breaking it again........ I can't even...I am so sorry

**Reply by spy16** [reply]
If you are serving the frontend files also from pocketbase binary in production, a neat little trick that might work if you are using Vite is to set the pocketbase host like this in `vite.config.ts`:

```typescript
{
   // other configs
   server: {
        proxy: {
            "/api" : "http://localhost:8090",
        }
   }
}
```

and the pocketbase client initialised as 

```typescript
const client = new Pocketbase("")
```

(you still have to set the redirect URL separately though)
